# Bài làm thêm 1 - Numpy



In [313]:
import numpy as np

In [314]:
#1. Tạo mảng ngẫu nhiên 10 phần tử số nguyên từ 10-29
np.random.seed(2)
arr1 =np.random.randint(10,30,size=10)
arr1

array([18, 25, 23, 18, 21, 28, 21, 18, 17, 12])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([18, 25, 23, 18, 21, 28, 21, 18, 17, 12])
</pre>

</details>

In [315]:
#2. Cho biết kiểu dữ liệu các phần tử của arr1
arr1.dtype


dtype('int32')

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>dtype('int32')
</pre>

</details>

In [316]:
#3. Cho biết dung lượng bộ nhớ của mảng arr1
arr1.nbytes


40

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>40
</pre>

</details>

In [317]:
#4. Tạo mảng arr2 là mảng arr1 nhưng có các giá trị chia hết cho 3 được thay bằng -1
arr2 = np.where(arr1 % 3==0,-1,arr1)
arr2


array([-1, 25, 23, -1, -1, 28, -1, -1, 17, -1])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([-1, 25, 23, -1, -1, 28, -1, -1, 17, -1])
</pre>

</details>

In [318]:
#5. Phát sinh mảng arr3 ngẫu nhiên 10 phần tử số nguyên từ 15-35
np.random.seed(3)
arr3 = np.random.randint(15,36,10)
print(arr3)

[25 18 23 15 34 25 26 24 25 21]


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([25, 18, 23, 15, 34, 25, 26, 24, 25, 21])
</pre>

</details>

In [319]:
#6.  Liệt kê ra các phần tử giống nhau trong arr1 và arr3
print(arr1)
print(arr3)
np.intersect1d(arr1,arr3)

[18 25 23 18 21 28 21 18 17 12]
[25 18 23 15 34 25 26 24 25 21]


array([18, 21, 23, 25])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([23])
</pre>

</details>

In [320]:
#7. Tạo mảng arr4 là mảng arr1 nhưng có 2 dòng, 5 cột 
arr4 = arr1.reshape(2,5)
arr4

array([[18, 25, 23, 18, 21],
       [28, 21, 18, 17, 12]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[18, 25, 23, 18, 21],
       [28, 21, 18, 17, 12]])
</pre>

</details>

In [321]:
#8. Hoán đổi 2 dòng của mảng arr4
np.flip(arr4,0)

array([[28, 21, 18, 17, 12],
       [18, 25, 23, 18, 21]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[28, 21, 18, 17, 12],
       [18, 25, 23, 18, 21]])
</pre>

</details>

In [322]:
#8. Hoán đổi cột 1 và cột 3 của mảng arr4

arr4 = arr4[:,[2,1,0,3,4]]
arr4

array([[23, 25, 18, 18, 21],
       [18, 21, 28, 17, 12]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>Trước khi hoán đổi: 
 [[18 25 23 18 21]
 [28 21 18 17 12]]
Sau khi hoán đổi: 
 [[23 25 18 18 21]
 [18 21 28 17 12]]
</pre>

</details>

In [323]:
#9. Tạo mảng arr5 kích thước 8x8 với giá trị bên trong là 0 và các giá trị biên là 1
arr5 = np.zeros([8,8])
arr5[:,0] = arr5[7,:] =arr5[0,:] =arr5[:,7] = 1
arr5

array([[1., 1., 1., 1., 1., 1., 1., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 0., 0., 0., 0., 0., 0., 1.],
       [1., 1., 1., 1., 1., 1., 1., 1.]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>[[1 1 1 1 1 1 1]
 [1 0 0 0 0 0 1]
 [1 0 0 0 0 0 1]
 [1 0 0 0 0 0 1]
 [1 0 0 0 0 0 1]
 [1 0 0 0 0 0 1]
 [1 1 1 1 1 1 1]]
</pre>

</details>

In [324]:
#10. Cộng thêm 3 vào các phần tử của mảng arr5
arr5 = np.add(arr5,3)
arr5

array([[4., 4., 4., 4., 4., 4., 4., 4.],
       [4., 3., 3., 3., 3., 3., 3., 4.],
       [4., 3., 3., 3., 3., 3., 3., 4.],
       [4., 3., 3., 3., 3., 3., 3., 4.],
       [4., 3., 3., 3., 3., 3., 3., 4.],
       [4., 3., 3., 3., 3., 3., 3., 4.],
       [4., 3., 3., 3., 3., 3., 3., 4.],
       [4., 4., 4., 4., 4., 4., 4., 4.]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[4, 4, 4, 4, 4, 4, 4],
       [4, 3, 3, 3, 3, 3, 4],
       [4, 3, 3, 3, 3, 3, 4],
       [4, 3, 3, 3, 3, 3, 4],
       [4, 3, 3, 3, 3, 3, 4],
       [4, 3, 3, 3, 3, 3, 4],
       [4, 4, 4, 4, 4, 4, 4]])
</pre>

</details>

In [325]:
#11. Tạo mảng arr6 là mảng 1 chiều của mảng arr5
arr6 = arr5.flatten()
arr6

array([4., 4., 4., 4., 4., 4., 4., 4., 4., 3., 3., 3., 3., 3., 3., 4., 4.,
       3., 3., 3., 3., 3., 3., 4., 4., 3., 3., 3., 3., 3., 3., 4., 4., 3.,
       3., 3., 3., 3., 3., 4., 4., 3., 3., 3., 3., 3., 3., 4., 4., 3., 3.,
       3., 3., 3., 3., 4., 4., 4., 4., 4., 4., 4., 4., 4.])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([4, 4, 4, 4, 4, 4, 4, 4, 3, 3, 3, 3, 3, 4, 4, 3, 3, 3, 3, 3, 4, 4,
       3, 3, 3, 3, 3, 4, 4, 3, 3, 3, 3, 3, 4, 4, 3, 3, 3, 3, 3, 4, 4, 4,
       4, 4, 4, 4, 4])
</pre>

</details>

In [326]:
#12. Tạo mảng arr7 là 5 phần tử ngẫu nhiên từ arr3
np.random.seed(8)
arr7 = np.random.choice(arr3,5)
arr7

array([15, 34, 18, 21, 25])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([15, 34, 18, 21, 25])
</pre>

</details>

#### Part 2 - Đọc dữ liệu và tính toán thống kê

In [327]:
#1. Đọc dữ liệu từ tập tin shark.tsv (dữ liệu số vụ cá mập tấn công các tháng của 3 quốc gia USA,Australia,South Africa)
# và đưa vào mảng shark (Gợi ý: sử dụng hàm np.genfromtxt)
shark = np.genfromtxt(fname="shark.tsv",delimiter="\t", skip_header=1,filling_values=-1)
shark


array([[ -1.,  28.,  94.,  68.],
       [ -1.,  27.,  78.,  32.],
       [ -1.,  66.,  63.,  34.],
       [ -1., 103.,  54.,  25.],
       [ -1., 106.,  21.,  25.],
       [ -1., 173.,  25.,  22.],
       [ -1., 232.,  25.,  28.],
       [ -1., 208.,  28.,  15.],
       [ -1., 182.,  35.,  17.],
       [ -1., 141.,  56.,  16.],
       [ -1.,  79.,  65.,  20.],
       [ -1.,  37.,  94.,  46.]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[ -1,  28,  94,  68],
       [ -1,  27,  78,  32],
       [ -1,  66,  63,  34],
       [ -1, 103,  54,  25],
       [ -1, 106,  21,  25],
       [ -1, 173,  25,  22],
       [ -1, 232,  25,  28],
       [ -1, 208,  28,  15],
       [ -1, 182,  35,  17],
       [ -1, 141,  56,  16],
       [ -1,  79,  65,  20],
       [ -1,  37,  94,  46]])
</pre>

</details>

In [328]:
#2. Cho biết số phần tử mỗi chiều của mảng
shark.shape


(12, 4)

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>(12, 4)
</pre>

</details>

In [329]:
#3. Thay cột 1 (giá trị -1) bằng giá trị tương ứng các tháng trong năm 1-12
mounth = np.arange(1,13)
shark[:,0] = mounth
shark

array([[  1.,  28.,  94.,  68.],
       [  2.,  27.,  78.,  32.],
       [  3.,  66.,  63.,  34.],
       [  4., 103.,  54.,  25.],
       [  5., 106.,  21.,  25.],
       [  6., 173.,  25.,  22.],
       [  7., 232.,  25.,  28.],
       [  8., 208.,  28.,  15.],
       [  9., 182.,  35.,  17.],
       [ 10., 141.,  56.,  16.],
       [ 11.,  79.,  65.,  20.],
       [ 12.,  37.,  94.,  46.]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[  1,  28,  94,  68],
       [  2,  27,  78,  32],
       [  3,  66,  63,  34],
       [  4, 103,  54,  25],
       [  5, 106,  21,  25],
       [  6, 173,  25,  22],
       [  7, 232,  25,  28],
       [  8, 208,  28,  15],
       [  9, 182,  35,  17],
       [ 10, 141,  56,  16],
       [ 11,  79,  65,  20],
       [ 12,  37,  94,  46]])
</pre>

</details>

In [330]:
#4. Hiển thị số vụ cá mập tấn công trung bình của 3 quốc gia
np.mean(shark[:,1:],axis=0)


array([115.16666667,  53.16666667,  29.        ])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([115.16666667,  53.16666667,  29.        ])
</pre>

</details>

In [332]:
#5. Hiển thị số vụ cá mập tấn công trung bình của mỗi tháng  
np.mean(shark[:,1:],axis=1)

array([63.33333333, 45.66666667, 54.33333333, 60.66666667, 50.66666667,
       73.33333333, 95.        , 83.66666667, 78.        , 71.        ,
       54.66666667, 59.        ])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([63.33333333, 45.66666667, 54.33333333, 60.66666667, 50.66666667,
       73.33333333, 95.        , 83.66666667, 78.        , 71.        ,
       54.66666667, 59.        ])
</pre>

</details>

In [353]:
#6. Tạo thêm cột tổng số vụ cá mập tấn công theo mỗi tháng
total = np.sum(shark[:,1:],1)
total.shape
np.insert(shark,len(shark[0]),total,axis=1)



array([[  1.,  28.,  94.,  68., 190.],
       [  2.,  27.,  78.,  32., 137.],
       [  3.,  66.,  63.,  34., 163.],
       [  4., 103.,  54.,  25., 182.],
       [  5., 106.,  21.,  25., 152.],
       [  6., 173.,  25.,  22., 220.],
       [  7., 232.,  25.,  28., 285.],
       [  8., 208.,  28.,  15., 251.],
       [  9., 182.,  35.,  17., 234.],
       [ 10., 141.,  56.,  16., 213.],
       [ 11.,  79.,  65.,  20., 164.],
       [ 12.,  37.,  94.,  46., 177.]])

<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[  1,  28,  94,  68, 190],
       [  2,  27,  78,  32, 137],
       [  3,  66,  63,  34, 163],
       [  4, 103,  54,  25, 182],
       [  5, 106,  21,  25, 152],
       [  6, 173,  25,  22, 220],
       [  7, 232,  25,  28, 285],
       [  8, 208,  28,  15, 251],
       [  9, 182,  35,  17, 234],
       [ 10, 141,  56,  16, 213],
       [ 11,  79,  65,  20, 164],
       [ 12,  37,  94,  46, 177]])
</pre>

</details>

In [ ]:
#7. Cho biết những tháng có tổng số vụ cá mập tấn công >200


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>array([[  6, 173,  25,  22, 220],
       [  7, 232,  25,  28, 285],
       [  8, 208,  28,  15, 251],
       [  9, 182,  35,  17, 234],
       [ 10, 141,  56,  16, 213]])
</pre>

</details>

In [ ]:
#8. Cho biết tổng số vụ cá mập tấn công nhiều nhất trong tháng là bao nhiêu


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>285
</pre>

</details>

In [ ]:
#9. Cho biết tháng nào có tổng số vụ cá mập tấn công nhiều nhất


<details>
  <summary>Nhấn vào đây để xem kết quả !</summary>
  
  <pre>Tháng có tổng số vụ cá mập tấn công nhiều nhất là tháng : 7
</pre>

</details>